src/CodeGen_Vulkan_Dev.cpp

```diff
@@ -0,0 +1,2924 @@
+        result_id = composite_id;
+    }
+    builder.update_id(result_id);
+}
+
+SpvId CodeGen_Vulkan_Dev::SPIRV_Emitter::map_type_to_pair(const Type &t) {
+    debug(2) << "CodeGen_Vulkan_Dev::SPIRV_Emitter::map_type_to_pair(): " << t << "\n";
+    SpvId base_type_id = builder.declare_type(t);
+    SpvBuilder::StructMemberTypes member_type_ids = {base_type_id, base_type_id};
+    const std::string struct_name = std::string("_struct_") + type_to_c_type(t, false, false) + std::string("_pair");
+    SpvId struct_type_id = builder.declare_struct(struct_name, member_type_ids);
+    return struct_type_id;
+}
+
+void CodeGen_Vulkan_Dev::SPIRV_Emitter::visit(const Variable *var) {
+    debug(2) << "CodeGen_Vulkan_Dev::SPIRV_Emitter::visit(Variable): " << var->type << " " << var->name << "\n";
+    SpvId variable_id = symbol_table.get(var->name).first;
+    user_assert(variable_id != SpvInvalidId) << "CodeGen_Vulkan_Dev::SPIRV_Emitter::visit(Variable): Invalid symbol name!\n";
+    builder.update_id(variable_id);
+}
+
+void CodeGen_Vulkan_Dev::SPIRV_Emitter::visit(const IntImm *imm) {
+    if (imm->type.bits() == 8) {
+        const int8_t value = (int8_t)(imm->value);
+        SpvId constant_id = builder.declare_constant(imm->type, &value);
+        builder.update_id(constant_id);
+    } else if (imm->type.bits() == 16) {
+        const int16_t value = (int16_t)(imm->value);
+        SpvId constant_id = builder.declare_constant(imm->type, &value);
+        builder.update_id(constant_id);
+    } else if (imm->type.bits() == 32) {
+        const int32_t value = (int32_t)(imm->value);
+        SpvId constant_id = builder.declare_constant(imm->type, &value);
+        builder.update_id(constant_id);
+    } else if (imm->type.bits() == 64) {
+        const int64_t value = (int64_t)(imm->value);
+        SpvId constant_id = builder.declare_constant(imm->type, &value);
+        builder.update_id(constant_id);
+    } else {
+        internal_error << "Vulkan backend currently only supports 8-bit, 16-bit, 32-bit or 64-bit signed integers!\n";
+    }
+}
+
+void CodeGen_Vulkan_Dev::SPIRV_Emitter::visit(const UIntImm *imm) {
+    if (imm->type.bits() == 8) {
+        const uint8_t value = (uint8_t)(imm->value);
+        SpvId constant_id = builder.declare_constant(imm->type, &value);
+        builder.update_id(constant_id);
+    } else if (imm->type.bits() == 16) {
+        const uint16_t value = (uint16_t)(imm->value);
+        SpvId constant_id = builder.declare_constant(imm->type, &value);
+        builder.update_id(constant_id);
+    } else if (imm->type.bits() == 32) {
+        const uint32_t value = (uint32_t)(imm->value);
+        SpvId constant_id = builder.declare_constant(imm->type, &value);
+        builder.update_id(constant_id);
+    } else if (imm->type.bits() == 64) {
+        const uint64_t value = (uint64_t)(imm->value);
+        SpvId constant_id = builder.declare_constant(imm->type, &value);
+        builder.update_id(constant_id);
+    } else {
```

Comment:
Seems like these could be simplified / shortened through a simple template. (and float too?)

Reply:
Agreed.  I'll clean this up a bit. Thx!